cirq-core/cirq/protocols/decompose_protocol.py

```diff
@@ -200,55 +249,14 @@ def decompose(
             "acceptable to keep."
         )
 
-    if preserve_structure:
-        return _decompose_preserving_structure(
-            val,
-            intercepting_decomposer=intercepting_decomposer,
-            fallback_decomposer=fallback_decomposer,
-            keep=keep,
-            on_stuck_raise=on_stuck_raise,
-        )
-
-    def try_op_decomposer(val: Any, decomposer: Optional[OpDecomposer]) -> DecomposeResult:
-        if decomposer is None or not isinstance(val, ops.Operation):
-            return None
-        return decomposer(val)
-
-    output = []
-    queue: List[Any] = [val]
-    while queue:
-        item = queue.pop(0)
```

Comment:
wouldn't replacing `queue.pop(0)` with `queue.pop()` and `queue[:0] = x` with `queue.extend(reversed(x))` do the trick ? 

dfs order is just stack order so renaming `queue` to `stack` and replacing the lines should do the job. Did I miss something?

Reply:
Yes, this would be sufficient for replacing the BFS order to a DFS order. But what we really is not just replacing the output ordering but also make sure that the items in the recursive OP-TREE are iterated upon in true DFS based ordering. So, in a follow up PR (since it requires a bunch of changes to the interfaces), I want the following test to pass:

```python
def test_decompose_recursive_dfs():
    mock_qm = mock.Mock()

    class RecursiveDecompose(cirq.Gate):
        def __init__(self, recurse: bool = True):
            self.recurse = recurse

        def _num_qubits_(self) -> int:
            return 1

        def _decompose_(self, qubits):
            mock_qm.qalloc(self.recurse)
            yield RecursiveDecompose(recurse=False).on(*qubits) if self.recurse else cirq.Z(*qubits)
            mock_qm.qfree(self.recurse)

    gate = RecursiveDecompose()
    _ = cirq.decompose(gate.on(cirq.NamedQubit("q")))
    expected_calls = [
        mock.call.qalloc(True),
        mock.call.qalloc(False),
        mock.call.qfree(False),
        mock.call.qfree(True),
    ]
    mock_qm.assert_has_calls(expected_calls, any_order=False)
```


If you simply use a stack instead of a queue but continue the with the iterative approach; in the first step of iteration you'll end up yielding all the elements of `RecursiveDecompose(True)._decompose_()` instead of getting only the first value from the generator and then recursively decomposing it before moving on to the next statement. 

If we get a list of all elements from `RecursiveDecompose(True)._decompose_()` in a single shot and insert them in a stack; the output ordering of decomposed operations would follow the DFS ordering but we would be processing all children of a node at once; instead of processing them one by one. 

We could potentially get around this constraint by adding more if/else blocks and checking whether If the operation on top of the stack is an OP-TREE; only yield the first value from the generator and then push the modified OP-TREE back in the stack along with the first child. Then pop the first child and continue to decompose. But I think the recursive implementation here is a lot more intuitive; so unless there are significant performance concerns with using the recursive approach; I'd say let's make the change to a recursive approach now and we can always optimize back to the iterative approach later. 


What do you think? I can also prototype the iterative approach directly if you have a strong preference.

Reply:
flattening the OP_TREE and pushing it to the stack in reverse order is what we need
```
stack.extend(reversed(cirq.flatten_to_ops(cirq.decompose_once(val))))
```

---

I'm okay either way. it's just that it seems that modifying the iterative approach is a lot less work

Reply:
As I highlighted in my previous comment, flattening the OP-TREE will not work in this case. My argument will become more clear in my next follow up PR. I'll tag you once it's out and we can come back to the iterative vs recursive discussion at that point.